Rewrite Python code in Go

itaysk opened the pull request:
This is the initial rewrite of the userspace code in Go. there's still some work to do, but I'd like to merge it now as it's stable, and continue incrementally improving with additional PRs. This PR should have feature parity with commit 431eaae7d4107cb6179ee9486f70d9435e44ebe3.

Notes:

- There are some minor UX changes. For example: timestamp formatting, args are printed inside brackets, new output flag, events flag accepts one flag at a time but can be defined multiple times.

- @simar7  correctly noted that consts names for the argtypes doesn't match Go conventions (which say consts are no different than vars). However, after reading about it a bit, and considering that those consts are borrowed from the C and Python code, I'd like to keep it as is (CAPS_SNAKE). This is also demonstrated in the Go codebase where they borrow const names from C.

- The `context` struct field names are based on the `context` struct from the C code. Unlike the discussion about consts names, here this is not borrowed but redefined, hence I think it should be following the Go naming contentions. It was a bit challenging following go's "initialism" convention: https://github.com/golang/go/wiki/CodeReviewComments#initialisms since the names of the context fields are all initials, and exported, so technically should be all upper cased, but that didn't feel right to me so I ignored this convention in this case and treated those fields as regular names. Hence some linter errors which I'm fine with that in this particular case.

- I'm still deliberating the way Syscalls/Sysevents are defined as map (it was done to make lookup simple)

- @yanivagman pointed to 2  gobpf related issues: how to set buffer size, and how to handle lost events. I'll address those in a future PR.

- Tests are missing, as they are in the python version. We'll continue with what @simar7 has started in another PR.

I'll open a GH issue for each item but adding as well.

simar7 commented on Makefile.old:
```diff
@@ -0,0 +1,2 @@
+test:
```
Could we this to the same Makefile but under a different target? Something like `python-test` for instance.

yanivagman commented on tracee/printer.go:
```diff
@@ -0,0 +1,62 @@
+	if err != nil {
+		fmt.Printf("error printing event: %v\n", err)
+	}
+	fmt.Printf("%s", string(eBytes))
```
In the python code we support a function named "get_events" which retreives the json output when needed.
We need to think of a way to mimic that here, as it is needed by ultrabox - maybe output to a file?

itaysk replied:
as discussed, ultrabox cannot import this module anymore as it's now golang; ultrabox will have to exec the process. so either read stdout or redirect output to a file and read it.

itaysk commented on tracee/argprinters.go:
```diff
@@ -0,0 +1,340 @@
+
+// PrintMknodMode prints the `mode` bitmask argument of the `mknod` syscall
+// http://man7.org/linux/man-pages/man7/inode.7.html
+func PrintMknodMode(mode uint32) string {
```
Thanks. I still want to "qualify" the otherwise generic argument name "mode". How about `PrintInodeMode`?

yanivagman replied:
Sounds good to me. Another option - PrintFileMode

itaysk commented on tracee/argprinters.go:
```diff
@@ -0,0 +1,340 @@
+// PrintMmapProt prints the `prot` bitmask argument of the `mmap` syscall
+// http://man7.org/linux/man-pages/man2/mmap.2.html
+// https://elixir.bootlin.com/linux/v5.5.3/source/include/uapi/asm-generic/mman-common.h#L10
+func PrintMmapProt(prot uint32) string {
```
How about `PrintMemoryProt` (memprot?)

yanivagman replied:
This can work

itaysk commented on tracee/tracee.go:
```diff
@@ -0,0 +1,487 @@
+	Tid     uint32   `json:"tid"`
+	Ppid    uint32   `json:"ppid"`
+	Uid     uint32   `json:"uid"`
+	MntId   uint32   `json:"mnt_id"`
```
I used the same names as in the `context_t` struct. will use the names from the python `print_event`

itaysk commented on tracee/tracee.go:
```diff
@@ -0,0 +1,487 @@
+	var err error
+
+	// validation
+	err = cfg.Validate()
```
The state of mind here is that tracee is a library. The part of creating the configuration is a distinct operation from running tracee, and at this point we can already fail early if the input is user provided invalid. The `New` function is the entry point for the library, and it should be able to care for itself and validate input regardless of if a user validated before. They might create the config manually instead of using the `NewConfig` helper, or change the config before passing it to `New`.

itaysk commented on tracee/tracee.go:
```diff
@@ -0,0 +1,487 @@
+	if err != nil {
+		return nil, fmt.Errorf("validation error: %v", err)
+	}
+	bpfFile := "./tracee/event_monitor_ebpf.c"
```
I don't think this should be a package level constant. In some point we might want to make it configurable or load an elf instead or some other thing. by making this data a function level local instead of a package level global I "hide" it from the library users, and reserve the right to change it's behavior in the future.

itaysk commented on tracee/tracee.go:
```diff
@@ -0,0 +1,487 @@
+	bpfConfig.Set(key, leaf)
+	binary.LittleEndian.PutUint32(key, uint32(CONFIG_DETECT_ORIG_SYSCALL))
+	binary.LittleEndian.PutUint32(leaf, boolToUInt32(t.config.DetectOriginalSyscall))
+	bpfConfig.Set(key, leaf)
```
the table is initialized in the bpf code, no? I just checked `-c` and it works

itaysk commented on tracee/tracee.go:
```diff
@@ -0,0 +1,487 @@
+	bpfConfig.Set(key, leaf)
+
+	eventsBPFTable := bpf.NewTable(t.bpfModule.TableId("events"), t.bpfModule)
+	t.eventsChannel = make(chan []byte, 1000)
```
no good reason. will do in the future

itaysk commented on tracee/printer.go:
```diff
@@ -0,0 +1,62 @@
+	fmt.Println()
+}
+
+func (p tableEventPrinter) Epilogue() {}
```
noted (out of scope)

itaysk commented on tracee/tracee.go:
```diff
@@ -0,0 +1,487 @@
+
+	eventsBPFTable := bpf.NewTable(t.bpfModule.TableId("events"), t.bpfModule)
+	t.eventsChannel = make(chan []byte, 1000)
+	t.bpfPerfMap, err = bpf.InitPerfMap(eventsBPFTable, t.eventsChannel)
```
will do in the future

itaysk commented on tracee/tracee.go:
```diff
@@ -0,0 +1,487 @@
+		dataBuff := bytes.NewBuffer(dataRaw)
+		ctx, err := readContextFromBuff(dataBuff)
+		if err != nil {
+			continue
```
noted (out of scope)

itaysk commented:
Thanks @yanivagman ! You really caught some important issues there. 
I've addressed most of your comments:
- Those that I just "resolved" I fixed. 
- Some I didn't fix and answered why. 
- The one about not printing unwanted events (left opened) need some more planning, will to that tomorrow. 
- Some I wrote down to add in another PR:
  - show events and lost/err event count https://github.com/aquasecurity/tracee/pull/47/files#r381269827
https://github.com/aquasecurity/tracee/pull/47/files#r381326662
  - consider events channel size
https://github.com/aquasecurity/tracee/pull/47/files#r381313817
  - configurable perf map size 
https://github.com/aquasecurity/tracee/pull/47/files#r381322134

yanivagman commented on tracee/tracee.go:
```diff
@@ -0,0 +1,489 @@
+	return []byte(tc.String()), nil
+}
+
+// contex struct contains common metadata that is collected for all types of events
```
contex->context

yanivagman commented on tracee/tracee.go:
```diff
@@ -0,0 +1,487 @@
+				continue
+			}
+		}
+		t.printer.Print(ctx, args)
```
What about this one?

itaysk replied:
As I wrote:
> The one about not printing unwanted events (left opened) need some more planning, will to that tomorrow.

I'll let you know once ready to re-review

itaysk commented:
@yanivagman in order to fix https://github.com/aquasecurity/tracee/pull/47#discussion_r381311011 with minimum performance penalty, I did some refactoring to how events and their ids are managed (I also took the opportunity to make some changes I wanted to).

I'm still unsure about the last commit https://github.com/aquasecurity/tracee/pull/47/commits/abf4e6e56eab40a80a6367677016c557b7222d80 which changes tracee to use event ids instead of strings internally - Interested in your opinion. If we don't like it we can revert it and fall back to https://github.com/aquasecurity/tracee/pull/47/commits/8383876cb9d3478d2c5be85a68a9a3997c89a87f which include all the other required changes.

yanivagman commented on main.go:
```diff
@@ -15,7 +15,7 @@ func main() {
 		Action: func(c *cli.Context) error {
 			if c.Bool("list") {
-				printList()
+				fmt.Println(tracee.EventsIDToName)
```
I think we should keep the separation between syscalls and other events when printing

itaysk replied:
fixed in 7b9a03445e99c03999c3fa4bc922063eeba87111

yanivagman commented on tracee/tracee.go:
```diff
@@ -254,13 +213,60 @@ func (t *Tracee) initBPF() error {
+	if !isEssential {
+		return true
+	}
+
```
If we reached this point, it means the event is an essential event.
As the number of the essential events is only 3, it is better to save a boolean for each of these ones once during init.
This will save the need to iterate through EventsToTrace which can be a big list

itaysk replied:
I have changed how shouldPrintEvent works altogether in https://github.com/aquasecurity/tracee/pull/47/commits/95dbdbe8586281a6ac4aaea089b4a15303b4f8b5. What do you think?

yanivagman replied:
Looks good, but I added a comment there

yanivagman commented on tracee/tracee.go:
```diff
@@ -487,3 +495,16 @@ func readArgFromBuff(dataBuff io.Reader) (interface{}, error) {
+func DedupStringSlice(in []string) []string {
+	tmpset := make(map[string]bool, len(in))
+	for _, e := range in {
+		tmpset[e] = true
```
This logic can be done in one iteration instead of 2:
if _, ok := tmpset[e]; !ok {
    out = append(out, e)
}

itaysk replied:
👍

yanivagman commented on tracee/consts.go:
```diff
@@ -393,10 +393,10 @@ func init() {
-	"do_exit",
-	"execve",
-	"execveat",
+var essentialEvents = []int32{
```
I don't really understand the reason for moving from string to int32 - can you please explain?

itaysk replied:
We just need to pick one type as the native type and I wasn't sure which one so I tried both. the fact that I tried first string and then int was just a coincidence. using string will not save us conversions, they will just move to other places.

yanivagman commented on tracee/tracee.go:
```diff
@@ -164,11 +164,24 @@ func (t *Tracee) initBPF() error {
+		} else {
+			eventsToTraceFinal = append(eventsToTraceFinal, e)
+		}
+		tmpset[e] = true
```
In the previous loop, tmpset contained only essential events.
This is being used in "if essential" above.
Here you add non essential events to tmpset.
So the assumption that tmpset contains only essential events is invalidated after one iteration.

itaysk replied:
there is not assumption that it contains only essential events, but I agree some of the var names like `essential` is misleading. There is another bug here where in the event that `eventsToTrace` contains duplicates, `essentialEvents` map will contain non-essential events. I was working under the assumption that the input is deduped, but I've addressed this concern in: 
https://github.com/aquasecurity/tracee/pull/47/commits/37b85b67c9d97608e2acfc1356767b10453011b8

yanivagman commented on tracee/tracee.go:
```diff
@@ -0,0 +1,500 @@
+		if err != nil {
+			return nil, fmt.Errorf("error reading syscall arg: %v", err)
+		}
+		if e, ok := EventsIDToName[sc]; ok {
```
This will not work with your recent changes!
We need to have a full map of syscall ids to syscall names

itaysk replied:
Why do you think this will not work?

itaysk replied:
got it, fixed in 33e946e72f0ba58500b440622c1a7653afe6ade4
